### ISSUE TITLE
Incorrect gradient from tfp.distributions.RelaxedOneHotCategorical

### ISSUE THREAD

**Issue by mmilosav** [tensorflow/probability#1132]
I am attempting to differentiate through `RelaxedOneHotCategorical`.  The gradients seem wrong: they are identically zero or sometimes tiny.  In contrast, the gradients form `ExpRelaxedOneHotCategorical` seem reasonable.  Given that `RelaxedOneHotCategorical` is the composition of `ExpRelaxedOneHotCategorical` and `exp_bijector.Exp()`, if the gradients of `ExpRelaxedOneHotCategorical` are or the order of unity, the gradients of `RelaxedOneHotCategorical` should also be of the order of unity.  Could it be that differentiation through the bijector fails?  The following code reproduces the issue on TF 2.3.1 and TFP 0.11.1. Please advise. Thank you.
```python
import tensorflow as tf
import tensorflow_probability as tfp

relaxed = tfp.layers.DistributionLambda(
    make_distribution_fn=lambda probs: tfp.distributions.RelaxedOneHotCategorical(probs=probs, temperature=1.))

exp_relaxed = tfp.layers.DistributionLambda(
    make_distribution_fn=lambda probs: tfp.distributions.ExpRelaxedOneHotCategorical(probs=probs, temperature=1.))

probs = tf.Variable([[0.5, 0.5]])

with tf.GradientTape() as tape:
    sample = tf.convert_to_tensor(exp_relaxed(probs))

print(tape.gradient(sample, probs).numpy())  # gradients are of the order of unity

with tf.GradientTape() as tape:
    sample = tf.convert_to_tensor(relaxed(probs))

# by chain rule gradient(relaxed, probs) == exp(exp_relaxed) * gradient(exp_relaxed, probs)
# so the gradient should also be of the order of unity
# but the following line unexpectedly prints either [[0., 0.]] or very small numbers!!!
print(tape.gradient(sample, probs).numpy())
```

**Comment by mmilosav** [comment]
I further notice that a related issue affects `tfp.distributions.Dirichlet` and `tfp.distributions.ExpRelaxedOneHotCategorical`.  Gradients from samples are correct, but gradients from samples transformed by a math function are not.  For both `Dirichlet` and `ExpRelaxedOneHotCategorical`, the following pseudocode **prints zero gradients**:
```python
def func(x):
     return tf.math.some_function(x)

param = tf.constant(...)

layer = tfp.layers.DistributionLambda(lambda param_: tfp.distributions.Distribution(param_))

with tf.GradientTape() as tape:
    tape.watch(param)
    func_of_sample = func(layer(param))

 print(tape.gradient(func_of_sample, param))
```
The same happens if `func(...)` is wrapped in `tf.keras.layers.Lambda`, and if the layers tracked in a subclassed `tf.keras.Model`.  

However for `Normal` and `MultivariateNormalDiag`, the gradients are correct and nonzero.

Is there something different in the implementations of {`Dirichlet`, `ExpRelaxedOneHotCategorical`} vs. {`Normal`, `MultivariateNormalDiag`} that ensures the correct backward pass behavior in the latter set, but not in the former?  Please advise. Cc. @brianwa84 .  Thank you.

**Comment by brianwa84** [comment]
cc @srvasude @axch 
My first thought was that the gradient wrt the sample is lost because the sampling happens implicitly, between keras layers.

But writing this out directly, I get pretty similar results (though not always zero).
Have you played with a wide range of temperature values?

@srvasude was the bug in gamma sample autodiff fixed by the 0.11 release?

**Comment by brianwa84** [comment]
btw, you can drop the keras code like so:
```python
probs = tf.Variable([[.5, .5]])
with tf.GradientTape(persistent=True) as tape:
  sample = tfp.distributions.ExpRelaxedOneHotCategorical(probs=probs, temperature=[.2, 1., 5]).sample()
  sample1 = tf.math.exp(sample)
  sample2 = tfp.distributions.RelaxedOneHotCategorical(probs=probs, temperature=[.2, 1., 5]).sample()
print(tape.gradient(sample, probs).numpy(),'\n')
print(sample1.numpy(), sample, tape.gradient(sample1, probs).numpy(),'\n')
print(sample2.numpy(), np.log(sample2), tape.gradient(sample2, probs).numpy(),'\n')
```

**Comment by mmilosav** [comment]
@brianwa84 Indeed the gradients are either zero or tiny, ~1e-7 or smaller, regardless of temperature.

Also, this affects straight up gradient of `Dirichlet`:
```python
alpha = tf.Variable([[2., 0.5]])

with tf.GradientTape() as tape:
    sample = tfp.distributions.Dirichlet(concentration=alpha, validate_args=True).sample()

print(sample.numpy(), tape.gradient(sample, alpha).numpy())  # prints zero or tiny gradient
```

**Comment by mmilosav** [comment]
@brianwa84 I apologize to have caused confusion.  I overlooked that "gradient" is the Jacobian **reduced along the output dimension**.  As such the "gradient" vanishes for distributions on the probability simplex.  This is the expected behavior.  Closing this issue.
